internal/ui/backup/progress.go

```diff
@@ -84,191 +82,165 @@ func NewProgress(printer ProgressPrinter, interval time.Duration) *Progress {
 // Run regularly updates the status lines. It should be called in a separate
 // goroutine.
 func (p *Progress) Run(ctx context.Context) {
-	var (
-		lastUpdate       time.Time
-		total, processed Counter
-		errors           uint
-		started          bool
-		currentFiles     = make(map[string]struct{})
-		secondsRemaining uint64
-	)
-
-	t := time.NewTicker(time.Second)
-	signalsCh := signals.GetProgressChannel()
-	defer t.Stop()
 	defer close(p.closed)
 	// Reset status when finished
 	defer p.printer.Reset()
 
+	if p.interval == 0 {
+		return
+	}
+
+	t := time.NewTicker(p.interval)
+	defer t.Stop()
+
+	signalsCh := signals.GetProgressChannel()
+
 	for {
-		forceUpdate := false
+		var now time.Time
 		select {
 		case <-ctx.Done():
 			return
-		case t, ok := <-p.totalCh:
-			if ok {
-				total = t
-				started = true
-			} else {
-				// scan has finished
-				p.totalCh = nil
-			}
-		case s := <-p.processedCh:
-			processed.Files += s.Files
-			processed.Dirs += s.Dirs
-			processed.Bytes += s.Bytes
-			started = true
-		case <-p.errCh:
-			errors++
-			started = true
-		case m := <-p.workerCh:
-			if m.done {
-				delete(currentFiles, m.filename)
-			} else {
-				currentFiles[m.filename] = struct{}{}
-			}
-		case <-t.C:
-			if !started {
-				continue
-			}
-
-			if p.totalCh == nil {
-				secs := float64(time.Since(p.start) / time.Second)
-				todo := float64(total.Bytes - processed.Bytes)
-				secondsRemaining = uint64(secs / float64(processed.Bytes) * todo)
-			}
+		case now = <-t.C:
 		case <-signalsCh:
-			forceUpdate = true
+			now = time.Now()
 		}
 
-		// limit update frequency
-		if !forceUpdate && (p.interval == 0 || time.Since(lastUpdate) < p.interval) {
+		p.mu.Lock()
+		state := p.state
+
+		if state&stateStarted == 0 {
+			p.mu.Unlock()
 			continue
 		}
-		lastUpdate = time.Now()
 
-		p.printer.Update(total, processed, errors, currentFiles, p.start, secondsRemaining)
+		var secondsRemaining uint64
+		if state&stateFinished != 0 {
```

Comment:
`stateFinished` is highly misleading, as it appears to complement `stateStarted`, but that is not the case. `scanFinished` would be better.